diff --git a/stream_fork.sv b/stream_fork.sv
index 7ae27c1a06..e27b6e4549 100644
--- a/stream_fork.sv
+++ b/stream_fork.sv
@@ -29,7 +29,7 @@ module stream_fork #(
     input  logic [N_OUP-1:0]    ready_i
 );
 
-    typedef enum logic {READY, WAIT} state_t;
+    typedef enum logic {READY, WAITING} state_t;
 
     logic [N_OUP-1:0]   oup_ready,
                         all_ones;
@@ -50,13 +50,13 @@ module stream_fork #(
                     end else begin
                         ready_o = 1'b0;
                         // Otherwise, wait for inputs that did not handshake yet.
-                        inp_state_d = WAIT;
+                        inp_state_d = WAITING;
                     end
                 end else begin
                     ready_o = 1'b0;
                 end
             end
-            WAIT: begin
+            WAITING: begin
                 if (valid_i && oup_ready == all_ones) begin
                     ready_o = 1'b1;
                     inp_state_d = READY;
@@ -94,14 +94,14 @@ module stream_fork #(
                         valid_o[i] = 1'b1;
                         if (ready_i[i]) begin   // Output handshake
                             if (!ready_o) begin     // No input handshake yet
-                                oup_state_d = WAIT;
+                                oup_state_d = WAITING;
                             end
                         end else begin          // No output handshake
                             oup_ready[i] = 1'b0;
                         end
                     end
                 end
-                WAIT: begin
+                WAITING: begin
                     if (valid_i && ready_o) begin   // Input handshake
                         oup_state_d = READY;
                     end
