diff --git a/riscv_compliance/rtl/ibex_riscv_compliance.sv b/riscv_compliance/rtl/ibex_riscv_compliance.sv
index 8d3669ef33..aa1c4628fe 100644
--- a/riscv_compliance/rtl/ibex_riscv_compliance.sv
+++ b/riscv_compliance/rtl/ibex_riscv_compliance.sv
@@ -208,7 +208,19 @@ module ibex_riscv_compliance (
       .alert_minor_o          (                     ),
       .alert_major_internal_o (                     ),
       .alert_major_bus_o      (                     ),
-      .core_sleep_o           (                     )
+      .core_sleep_o           (                     ),
+
+      .lockstep_cmp_en_o         (                     ),
+
+      .data_req_shadow_o         (                     ),
+      .data_we_shadow_o          (                     ),
+      .data_be_shadow_o          (                     ),
+      .data_addr_shadow_o        (                     ),
+      .data_wdata_shadow_o       (                     ),
+      .data_wdata_intg_shadow_o  (                     ),
+
+      .instr_req_shadow_o        (                     ),
+      .instr_addr_shadow_o       (                     )
     );
 
   // SRAM block for instruction and data storage
diff --git a/uvm/core_ibex/tb/core_ibex_tb_top.sv b/uvm/core_ibex/tb/core_ibex_tb_top.sv
index 8178caecbc..b96137df0a 100644
--- a/uvm/core_ibex/tb/core_ibex_tb_top.sv
+++ b/uvm/core_ibex/tb/core_ibex_tb_top.sv
@@ -154,7 +154,18 @@ module core_ibex_tb_top;
     .alert_minor_o          (dut_if.alert_minor         ),
     .alert_major_internal_o (dut_if.alert_major_internal),
     .alert_major_bus_o      (dut_if.alert_major_bus     ),
-    .core_sleep_o           (dut_if.core_sleep          )
+    .core_sleep_o           (dut_if.core_sleep          ),
+
+    .lockstep_cmp_en_o         (                           ),
+    .data_req_shadow_o         (                           ),
+    .data_we_shadow_o          (                           ),
+    .data_be_shadow_o          (                           ),
+    .data_addr_shadow_o        (                           ),
+    .data_wdata_shadow_o       (                           ),
+    .data_wdata_intg_shadow_o  (                           ),
+
+    .instr_req_shadow_o        (                           ),
+    .instr_addr_shadow_o       (                           )
   );
 
   `define IBEX_RF_PATH core_ibex_tb_top.dut.u_ibex_top.gen_regfile_ff.register_file_i
