From 0003fac5417788dd8aaeb6eebcd767049d3a5859 Mon Sep 17 00:00:00 2001
From: Timothy Chen <timothytim@google.com>
Date: Wed, 11 May 2022 14:54:30 -0700
Subject: [PATCH 2/2] [PATCH] Minor lint fix

Signed-off-by: Timothy Chen <timothytim@google.com>
---
 src/dmi_jtag.sv | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/src/dmi_jtag.sv b/src/dmi_jtag.sv
index e040a3a..048dbff 100644
--- a/src/dmi_jtag.sv
+++ b/src/dmi_jtag.sv
@@ -59,6 +59,7 @@ module dmi_jtag #(
   logic tdi;
 
   logic dtmcs_select;
+  dm::dtmcs_t dtmcs_d, dtmcs_q;
 
   assign dmi_clear = jtag_dmi_clear || (dtmcs_select && update && dtmcs_q.dmihardreset);
 
@@ -66,8 +67,6 @@ module dmi_jtag #(
   // Debug Module Control and Status
   // -------------------------------
 
-  dm::dtmcs_t dtmcs_d, dtmcs_q;
-
   always_comb begin
     dtmcs_d = dtmcs_q;
     if (capture) begin
-- 
2.36.1.124.g0e6072fb45-goog

