diff --git a/cryptoki/src/mechanism/mod.rs b/cryptoki/src/mechanism/mod.rs
index 884ffcd..0000000 100644
--- a/cryptoki/src/mechanism/mod.rs
+++ b/cryptoki/src/mechanism/mod.rs
@@ -822,8 +822,12 @@ impl TryFrom<CK_MECHANISM_TYPE> for MechanismType {
             CKM_SP800_108_FEEDBACK_KDF => Ok(MechanismType::SP800_108_FEEDBACK_KDF),
             CKM_SP800_108_DOUBLE_PIPELINE_KDF => Ok(MechanismType::SP800_108_DOUBLE_PIPELINE_KDF),
             other => {
-                error!("Mechanism type {} is not supported.", other);
-                Err(Error::NotSupported)
+                if mechanism_type >= CKM_VENDOR_DEFINED {
+                    Ok(MechanismType { val: other })
+                } else {
+                    error!("Mechanism type {} is not supported.", other);
+                    Err(Error::NotSupported)
+                }
             }
         }
     }
