diff --git a/bindgen/private/bindgen.bzl b/bindgen/private/bindgen.bzl
--- bindgen/private/bindgen.bzl
+++ bindgen/private/bindgen.bzl
@@ -152,11 +152,6 @@ def _generate_cc_link_build_info(ctx, cc_lib):
                 linker_search_paths.append(lib.pic_static_library.dirname)
                 compile_data.append(lib.pic_static_library)
 
-    if not compile_data:
-        fail("No static libraries found in {}".format(
-            cc_lib.label,
-        ))
-
     rustc_flags_file = ctx.actions.declare_file("{}.rustc_flags".format(ctx.label.name))
     ctx.actions.write(
         output = rustc_flags_file,
