diff --git a/dv_utils/dv_utils.core b/dv_utils/dv_utils.core
index 0dba0235..801e84f6 100644
--- a/dv_utils.core
+++ b/dv_utils.core
@@ -11,7 +11,7 @@ filesets:
       - lowrisc:dv:dv_macros
       - lowrisc:dv:dv_fcov_macros
       - lowrisc:dv:common_ifs
       - lowrisc:prim:assert:0.1
-      - lowrisc:opentitan:bus_params_pkg
+      - lowrisc:ibex:bus_params_pkg
       - lowrisc:dv:str_utils
       - lowrisc:dv:dv_test_status
     files:
