diff --git a/ibex_trvk.sv b/ibex_trvk.sv
index 81c665e4e5..9963f9d582 100644
--- a/ibex_trvk.sv
+++ b/ibex_trvk.sv
@@ -127,7 +127,7 @@ module ibex_trvk #(
   // Base (address) calculation
   cap_meta_t    cap_meta;
   logic         unused_cap_meta;
-  logic         is_sealing_cap;
+  logic         cap_is_sealing;
   logic [32:0]  cap_base_33;
   logic         unused_cap_base_33;
   logic [31:0]  cap_base;
@@ -292,6 +292,10 @@ module ibex_trvk #(
     .err_o   ()
   );
 
+  // Only consumed by an assertion below
+  logic    unused_downstream_rsp_wready;
+  assign unused_downstream_rsp_wready = ^downstream_rsp_wready;
+
 
   ////////////////////////////////
   // Bitmap Address Calculation //
@@ -308,7 +312,7 @@ module ibex_trvk #(
   assign unused_cap_meta = ^{cap_meta.otype, cap_meta.cperms[5]};
 
   // Check if cap is sealing cap
-  assign is_sealing_cap = ibex_cheriot_pkg::is_sealing_cap(cap_meta.cperms[4:0]);
+  assign cap_is_sealing = ibex_cheriot_pkg::is_sealing_cap(cap_meta.cperms[4:0]);
 
   // Extract the middle field from the pointer, bounds depend on exponent, width fixed
   assign addr_mid = cheriot_bot_t'(ptr_storage_q >> cap_meta.exponent);
@@ -349,7 +353,7 @@ module ibex_trvk #(
 
   // We have loaded valid capability pointer, now we see valid metadata, not a sealing cap,
   // and are pointing into the revocation bitmap
-  assign revbm_req_required = !is_sealing_cap          && // Not sealing cap
+  assign revbm_req_required = !cap_is_sealing          && // Not sealing cap
                               ptr_storage_valid_q      && // The pointer stored is valid
                               downstream_rsp_out.tag   && // We are looking at a capability
                               downstream_rsp_out_valid && // The stored response is valid
