diff --git a/ibex_register_file_ff.sv b/ibex_register_file_ff.sv
index 53c89e2d34..f94f9550ad 100644
--- a/ibex_register_file_ff.sv
+++ b/ibex_register_file_ff.sv
@@ -14,8 +14,6 @@ module ibex_register_file_ff #(
   parameter bit                   RV32E             = 0,
   parameter int unsigned          DataWidth         = 32,
   parameter bit                   DummyInstructions = 0,
-  parameter bit                   WrenCheck         = 0,
-  parameter bit                   RdataMuxCheck     = 0,
   parameter logic [DataWidth-1:0] WordZeroVal       = '0
 ) (
   // Clock and Reset
@@ -38,10 +36,7 @@ module ibex_register_file_ff #(
   // Write port W1
   input  logic [4:0]           waddr_a_i,
   input  logic [DataWidth-1:0] wdata_a_i,
-  input  logic                 we_a_i,
-
-  // This indicates whether spurious WE or non-one-hot encoded raddr are detected.
-  output logic                 err_o
+  input  logic                 we_a_i
 );
 
   localparam int unsigned ADDR_WIDTH = RV32E ? 4 : 5;
@@ -50,44 +45,14 @@ module ibex_register_file_ff #(
   logic [DataWidth-1:0] rf_reg   [NUM_WORDS];
   logic [NUM_WORDS-1:0] we_a_dec;
 
-  logic oh_raddr_a_err, oh_raddr_b_err, oh_we_err;
-
   always_comb begin : we_a_decoder
     for (int unsigned i = 0; i < NUM_WORDS; i++) begin
       we_a_dec[i] = (waddr_a_i == 5'(i)) ? we_a_i : 1'b0;
     end
   end
 
-  // SEC_CM: DATA_REG_SW.GLITCH_DETECT
-  // This checks for spurious WE strobes on the regfile.
-  if (WrenCheck) begin : gen_wren_check
-    // Buffer the decoded write enable bits so that the checker
-    // is not optimized into the address decoding logic.
-    logic [NUM_WORDS-1:0] we_a_dec_buf;
-    prim_buf #(
-      .Width(NUM_WORDS)
-    ) u_prim_buf (
-      .in_i(we_a_dec),
-      .out_o(we_a_dec_buf)
-    );
-
-    prim_onehot_check #(
-      .AddrWidth(ADDR_WIDTH),
-      .AddrCheck(1),
-      .EnableCheck(1)
-    ) u_prim_onehot_check (
-      .clk_i,
-      .rst_ni,
-      .oh_i(we_a_dec_buf),
-      .addr_i(waddr_a_i),
-      .en_i(we_a_i),
-      .err_o(oh_we_err)
-    );
-  end else begin : gen_no_wren_check
-    logic unused_strobe;
-    assign unused_strobe = we_a_dec[0]; // this is never read from in this case
-    assign oh_we_err = 1'b0;
-  end
+  logic unused_strobe;
+  assign unused_strobe = we_a_dec[0]; // this is never read from in this case
 
   // No flops for R0 as it's hard-wired to 0
   for (genvar i = 1; i < NUM_WORDS; i++) begin : g_rf_flops
@@ -133,106 +98,8 @@ module ibex_register_file_ff #(
     assign rf_reg[0] = WordZeroVal;
   end
 
-  if (RdataMuxCheck) begin : gen_rdata_mux_check
-    // Encode raddr_a/b into one-hot encoded signals.
-    logic [NUM_WORDS-1:0] raddr_onehot_a, raddr_onehot_b;
-    logic [NUM_WORDS-1:0] raddr_onehot_a_buf, raddr_onehot_b_buf;
-    prim_onehot_enc #(
-      .OneHotWidth(NUM_WORDS)
-    ) u_prim_onehot_enc_raddr_a (
-      .in_i  (raddr_a_i),
-      .en_i  (1'b1),
-      .out_o (raddr_onehot_a)
-    );
-
-    prim_onehot_enc #(
-      .OneHotWidth(NUM_WORDS)
-    ) u_prim_onehot_enc_raddr_b (
-      .in_i  (raddr_b_i),
-      .en_i  (1'b1),
-      .out_o (raddr_onehot_b)
-    );
-
-    // Buffer the one-hot encoded signals so that the checkers
-    // are not optimized.
-    prim_buf #(
-      .Width(NUM_WORDS)
-    ) u_prim_buf_raddr_a (
-      .in_i (raddr_onehot_a),
-      .out_o(raddr_onehot_a_buf)
-    );
-
-    prim_buf #(
-      .Width(NUM_WORDS)
-    ) u_prim_buf_raddr_b (
-      .in_i (raddr_onehot_b),
-      .out_o(raddr_onehot_b_buf)
-    );
-
-    // SEC_CM: DATA_REG_SW.GLITCH_DETECT
-    // Check the one-hot encoded signals for glitches.
-    prim_onehot_check #(
-      .AddrWidth(ADDR_WIDTH),
-      .OneHotWidth(NUM_WORDS),
-      .AddrCheck(1),
-      // When AddrCheck=1 also EnableCheck needs to be 1.
-      .EnableCheck(1)
-    ) u_prim_onehot_check_raddr_a (
-      .clk_i,
-      .rst_ni,
-      .oh_i   (raddr_onehot_a_buf),
-      .addr_i (raddr_a_i),
-      // Set enable=1 as address is always valid.
-      .en_i   (1'b1),
-      .err_o  (oh_raddr_a_err)
-    );
-
-    prim_onehot_check #(
-      .AddrWidth(ADDR_WIDTH),
-      .OneHotWidth(NUM_WORDS),
-      .AddrCheck(1),
-      // When AddrCheck=1 also EnableCheck needs to be 1.
-      .EnableCheck(1)
-    ) u_prim_onehot_check_raddr_b (
-      .clk_i,
-      .rst_ni,
-      .oh_i   (raddr_onehot_b_buf),
-      .addr_i (raddr_b_i),
-      // Set enable=1 as address is always valid.
-      .en_i   (1'b1),
-      .err_o  (oh_raddr_b_err)
-    );
-
-    // MUX register to rdata_a/b_o according to raddr_a/b_onehot.
-    prim_onehot_mux  #(
-      .Width(DataWidth),
-      .Inputs(NUM_WORDS)
-    ) u_rdata_a_mux (
-      .clk_i,
-      .rst_ni,
-      .in_i  (rf_reg),
-      .sel_i (raddr_onehot_a),
-      .out_o (rdata_a_o)
-    );
-
-    prim_onehot_mux  #(
-      .Width(DataWidth),
-      .Inputs(NUM_WORDS)
-    ) u_rdata_b_mux (
-      .clk_i,
-      .rst_ni,
-      .in_i  (rf_reg),
-      .sel_i (raddr_onehot_b),
-      .out_o (rdata_b_o)
-    );
-  end else begin : gen_no_rdata_mux_check
-    assign rdata_a_o = rf_reg[raddr_a_i];
-    assign rdata_b_o = rf_reg[raddr_b_i];
-    assign oh_raddr_a_err = 1'b0;
-    assign oh_raddr_b_err = 1'b0;
-  end
-
-  assign err_o = oh_raddr_a_err || oh_raddr_b_err || oh_we_err;
+  assign rdata_a_o = rf_reg[raddr_a_i];
+  assign rdata_b_o = rf_reg[raddr_b_i];
 
   // Signal not used in FF register file
   logic unused_test_en;
diff --git a/ibex_register_file_fpga.sv b/ibex_register_file_fpga.sv
index 73e0afab2f..fcabad15c0 100644
--- a/ibex_register_file_fpga.sv
+++ b/ibex_register_file_fpga.sv
@@ -15,8 +15,6 @@ module ibex_register_file_fpga #(
     parameter bit                   RV32E             = 0,
     parameter int unsigned          DataWidth         = 32,
     parameter bit                   DummyInstructions = 0,
-    parameter bit                   WrenCheck         = 0,
-    parameter bit                   RdataMuxCheck     = 0,
     parameter logic [DataWidth-1:0] WordZeroVal       = '0
 ) (
   // Clock and Reset
@@ -36,10 +34,7 @@ module ibex_register_file_fpga #(
   // Write port W1
   input  logic [          4:0] waddr_a_i,
   input  logic [DataWidth-1:0] wdata_a_i,
-  input  logic                 we_a_i,
-
-  // This indicates whether spurious WE or non-one-hot encoded raddr are detected.
-  output logic                 err_o
+  input  logic                 we_a_i
 );
 
   localparam int ADDR_WIDTH = RV32E ? 4 : 5;
@@ -50,125 +45,12 @@ module ibex_register_file_fpga #(
 
   logic [DataWidth-1:0] mem_o_a, mem_o_b;
 
-  // WE strobe and one-hot encoded raddr alert.
-  logic oh_raddr_a_err, oh_raddr_b_err, oh_we_err;
-  assign err_o = oh_raddr_a_err || oh_raddr_b_err || oh_we_err;
-
-  if (RdataMuxCheck) begin : gen_rdata_mux_check
-    // Encode raddr_a/b into one-hot encoded signals.
-    logic [NUM_WORDS-1:0] raddr_onehot_a, raddr_onehot_b;
-    logic [NUM_WORDS-1:0] raddr_onehot_a_buf, raddr_onehot_b_buf;
-    prim_onehot_enc #(
-      .OneHotWidth(NUM_WORDS)
-    ) u_prim_onehot_enc_raddr_a (
-      .in_i  (raddr_a_i),
-      .en_i  (1'b1),
-      .out_o (raddr_onehot_a)
-    );
-
-    prim_onehot_enc #(
-      .OneHotWidth(NUM_WORDS)
-    ) u_prim_onehot_enc_raddr_b (
-      .in_i  (raddr_b_i),
-      .en_i  (1'b1),
-      .out_o (raddr_onehot_b)
-    );
-
-    // Buffer the one-hot encoded signals so that the checkers
-    // are not optimized.
-    prim_buf #(
-      .Width(NUM_WORDS)
-    ) u_prim_buf_raddr_a (
-      .in_i (raddr_onehot_a),
-      .out_o(raddr_onehot_a_buf)
-    );
-
-    prim_buf #(
-      .Width(NUM_WORDS)
-    ) u_prim_buf_raddr_b (
-      .in_i (raddr_onehot_b),
-      .out_o(raddr_onehot_b_buf)
-    );
-
-    // SEC_CM: DATA_REG_SW.GLITCH_DETECT
-    // Check the one-hot encoded signals for glitches.
-    prim_onehot_check #(
-      .AddrWidth(ADDR_WIDTH),
-      .OneHotWidth(NUM_WORDS),
-      .AddrCheck(1),
-      // When AddrCheck=1 also EnableCheck needs to be 1.
-      .EnableCheck(1)
-    ) u_prim_onehot_check_raddr_a (
-      .clk_i,
-      .rst_ni,
-      .oh_i   (raddr_onehot_a_buf),
-      .addr_i (raddr_a_i),
-      // Set enable=1 as address is always valid.
-      .en_i   (1'b1),
-      .err_o  (oh_raddr_a_err)
-    );
-
-    prim_onehot_check #(
-      .AddrWidth(ADDR_WIDTH),
-      .OneHotWidth(NUM_WORDS),
-      .AddrCheck(1),
-      // When AddrCheck=1 also EnableCheck needs to be 1.
-      .EnableCheck(1)
-    ) u_prim_onehot_check_raddr_b (
-      .clk_i,
-      .rst_ni,
-      .oh_i   (raddr_onehot_b_buf),
-      .addr_i (raddr_b_i),
-      // Set enable=1 as address is always valid.
-      .en_i   (1'b1),
-      .err_o  (oh_raddr_b_err)
-    );
-
-    // MUX register to rdata_a/b_o according to raddr_a/b_onehot.
-    prim_onehot_mux  #(
-      .Width(DataWidth),
-      .Inputs(NUM_WORDS)
-    ) u_rdata_a_mux (
-      .clk_i,
-      .rst_ni,
-      .in_i  (mem),
-      .sel_i (raddr_onehot_a),
-      .out_o (mem_o_a)
-    );
-
-    prim_onehot_mux  #(
-      .Width(DataWidth),
-      .Inputs(NUM_WORDS)
-    ) u_rdata_b_mux (
-      .clk_i,
-      .rst_ni,
-      .in_i  (mem),
-      .sel_i (raddr_onehot_b),
-      .out_o (mem_o_b)
-    );
-
-    assign rdata_a_o = (raddr_a_i == '0) ? WordZeroVal : mem_o_a;
-    assign rdata_b_o = (raddr_b_i == '0) ? WordZeroVal : mem_o_b;
-  end else begin : gen_no_rdata_mux_check
-    // async_read a
-    assign rdata_a_o = (raddr_a_i == '0) ? WordZeroVal : mem[raddr_a_i];
-
-    // async_read b
-    assign rdata_b_o = (raddr_b_i == '0) ? WordZeroVal : mem[raddr_b_i];
-  end
+  assign rdata_a_o = (raddr_a_i == '0) ? WordZeroVal : mem[raddr_a_i];
+  assign rdata_b_o = (raddr_b_i == '0) ? WordZeroVal : mem[raddr_b_i];
 
   // we select
   assign we = (waddr_a_i == '0) ? 1'b0 : we_a_i;
 
-  // SEC_CM: DATA_REG_SW.GLITCH_DETECT
-  // This checks for spurious WE strobes on the regfile.
-  if (WrenCheck) begin : gen_wren_check
-    // Since the FPGA uses a memory macro, there is only one write-enable strobe to check.
-    assign oh_we_err = we && !we_a_i;
-  end else begin : gen_no_wren_check
-    assign oh_we_err = 1'b0;
-  end
-
   // Note that the SystemVerilog LRM requires variables on the LHS of assignments within
   // "always_ff" to not be written to by any other process. However, to enable the initialization
   // of the inferred RAM32M primitives with non-zero values, below "initial" procedure is needed.
diff --git a/ibex_register_file_latch.sv b/ibex_register_file_latch.sv
index 4c295e2ed1..b0df830e08 100644
--- a/ibex_register_file_latch.sv
+++ b/ibex_register_file_latch.sv
@@ -15,8 +15,6 @@ module ibex_register_file_latch #(
   parameter bit                   RV32E             = 0,
   parameter int unsigned          DataWidth         = 32,
   parameter bit                   DummyInstructions = 0,
-  parameter bit                   WrenCheck         = 0,
-  parameter bit                   RdataMuxCheck     = 0,
   parameter logic [DataWidth-1:0] WordZeroVal       = '0
 ) (
   // Clock and Reset
@@ -38,10 +36,7 @@ module ibex_register_file_latch #(
   // Write port W1
   input  logic [4:0]           waddr_a_i,
   input  logic [DataWidth-1:0] wdata_a_i,
-  input  logic                 we_a_i,
-
-  // This indicates whether spurious WE or non-one-hot encoded raddr are detected.
-  output logic                 err_o
+  input  logic                 we_a_i
 );
 
   localparam int unsigned ADDR_WIDTH = RV32E ? 4 : 5;
@@ -51,8 +46,6 @@ module ibex_register_file_latch #(
 
   logic [NUM_WORDS-1:0] waddr_onehot_a;
 
-  logic oh_raddr_a_err, oh_raddr_b_err, oh_we_err;
-
   logic [NUM_WORDS-1:1] mem_clocks;
   logic [DataWidth-1:0] wdata_a_q;
 
@@ -65,109 +58,11 @@ module ibex_register_file_latch #(
 
   logic clk_int;
 
-  assign err_o = oh_raddr_a_err || oh_raddr_b_err || oh_we_err;
-
   //////////
   // READ //
   //////////
-  if (RdataMuxCheck) begin : gen_rdata_mux_check
-    // Encode raddr_a/b into one-hot encoded signals.
-    logic [NUM_WORDS-1:0] raddr_onehot_a, raddr_onehot_b;
-    logic [NUM_WORDS-1:0] raddr_onehot_a_buf, raddr_onehot_b_buf;
-    prim_onehot_enc #(
-      .OneHotWidth(NUM_WORDS)
-    ) u_prim_onehot_enc_raddr_a (
-      .in_i  (raddr_a_int),
-      .en_i  (1'b1),
-      .out_o (raddr_onehot_a)
-    );
-
-    prim_onehot_enc #(
-      .OneHotWidth(NUM_WORDS)
-    ) u_prim_onehot_enc_raddr_b (
-      .in_i  (raddr_b_int),
-      .en_i  (1'b1),
-      .out_o (raddr_onehot_b)
-    );
-
-    // Buffer the one-hot encoded signals so that the checkers
-    // are not optimized.
-    prim_buf #(
-      .Width(NUM_WORDS)
-    ) u_prim_buf_raddr_a (
-      .in_i(raddr_onehot_a),
-      .out_o(raddr_onehot_a_buf)
-    );
-
-    prim_buf #(
-      .Width(NUM_WORDS)
-    ) u_prim_buf_raddr_b (
-      .in_i(raddr_onehot_b),
-      .out_o(raddr_onehot_b_buf)
-    );
-
-    // SEC_CM: DATA_REG_SW.GLITCH_DETECT
-    // Check the one-hot encoded signals for glitches.
-    prim_onehot_check #(
-      .AddrWidth(ADDR_WIDTH),
-      .OneHotWidth(NUM_WORDS),
-      .AddrCheck(1),
-      // When AddrCheck=1 also EnableCheck needs to be 1.
-      .EnableCheck(1)
-    ) u_prim_onehot_check_raddr_a (
-      .clk_i,
-      .rst_ni,
-      .oh_i   (raddr_onehot_a_buf),
-      .addr_i (raddr_b_int),
-      // Set enable=1 as address is always valid.
-      .en_i   (1'b1),
-      .err_o  (oh_raddr_a_err)
-    );
-
-    prim_onehot_check #(
-      .AddrWidth(ADDR_WIDTH),
-      .OneHotWidth(NUM_WORDS),
-      .AddrCheck(1),
-      // When AddrCheck=1 also EnableCheck needs to be 1.
-      .EnableCheck(1)
-    ) u_prim_onehot_check_raddr_b (
-      .clk_i,
-      .rst_ni,
-      .oh_i   (raddr_onehot_b_buf),
-      .addr_i (raddr_b_int),
-      // Set enable=1 as address is always valid.
-      .en_i   (1'b1),
-      .err_o  (oh_raddr_b_err)
-    );
-
-    // MUX register to rdata_a/b_o according to raddr_a/b_onehot.
-    prim_onehot_mux  #(
-      .Width(DataWidth),
-      .Inputs(NUM_WORDS)
-    ) u_rdata_a_mux (
-      .clk_i,
-      .rst_ni,
-      .in_i  (mem),
-      .sel_i (raddr_onehot_a),
-      .out_o (rdata_a_o)
-    );
-
-    prim_onehot_mux  #(
-      .Width(DataWidth),
-      .Inputs(NUM_WORDS)
-    ) u_rdata_b_mux (
-      .clk_i,
-      .rst_ni,
-      .in_i  (mem),
-      .sel_i (raddr_onehot_b),
-      .out_o (rdata_b_o)
-    );
-  end else begin : gen_no_rdata_mux_check
-    assign rdata_a_o = mem[raddr_a_int];
-    assign rdata_b_o = mem[raddr_b_int];
-    assign oh_raddr_a_err = 1'b0;
-    assign oh_raddr_b_err = 1'b0;
-  end
+  assign rdata_a_o = mem[raddr_a_int];
+  assign rdata_b_o = mem[raddr_b_int];
 
   ///////////
   // WRITE //
@@ -203,36 +98,8 @@ module ibex_register_file_latch #(
     end
   end
 
-  // SEC_CM: DATA_REG_SW.GLITCH_DETECT
-  // This checks for spurious WE strobes on the regfile.
-  if (WrenCheck) begin : gen_wren_check
-    // Buffer the decoded write enable bits so that the checker
-    // is not optimized into the address decoding logic.
-    logic [NUM_WORDS-1:0] waddr_onehot_a_buf;
-    prim_buf #(
-      .Width(NUM_WORDS)
-    ) u_prim_buf (
-      .in_i(waddr_onehot_a),
-      .out_o(waddr_onehot_a_buf)
-    );
-
-    prim_onehot_check #(
-      .AddrWidth(ADDR_WIDTH),
-      .AddrCheck(1),
-      .EnableCheck(1)
-    ) u_prim_onehot_check (
-      .clk_i,
-      .rst_ni,
-      .oh_i(waddr_onehot_a_buf),
-      .addr_i(waddr_a_i),
-      .en_i(we_a_i),
-      .err_o(oh_we_err)
-    );
-  end else begin : gen_no_wren_check
-    logic unused_strobe;
-    assign unused_strobe = waddr_onehot_a[0]; // this is never read from in this case
-    assign oh_we_err = 1'b0;
-  end
+  logic unused_strobe;
+  assign unused_strobe = waddr_onehot_a[0]; // this is never read from in this case
 
   // Individual clock gating (if integrated clock-gating cells are available)
   for (genvar x = 1; x < NUM_WORDS; x++) begin : gen_cg_word_iter
diff --git a/ibex_top.sv b/ibex_top.sv
index 993736a808..d83b9dad96 100644
--- a/ibex_top.sv
+++ b/ibex_top.sv
@@ -146,8 +146,6 @@ module ibex_top import ibex_pkg::*; #(
   localparam bit          ResetAll              = Lockstep;
   localparam bit          DummyInstructions     = SecureIbex;
   localparam bit          RegFileECC            = SecureIbex;
-  localparam bit          RegFileWrenCheck      = SecureIbex;
-  localparam bit          RegFileRdataMuxCheck  = SecureIbex;
   localparam int unsigned RegFileDataWidth      = RegFileECC ? 32 + 7 : 32;
   localparam bit          MemECC                = SecureIbex;
   localparam int unsigned MemDataWidth          = MemECC ? 32 + 7 : 32;
@@ -416,15 +414,11 @@ module ibex_top import ibex_pkg::*; #(
   // Register file Instantiation //
   /////////////////////////////////
 
-  logic rf_alert_major_internal;
   if (RegFile == RegFileFF) begin : gen_regfile_ff
     ibex_register_file_ff #(
       .RV32E            (RV32E),
       .DataWidth        (RegFileDataWidth),
       .DummyInstructions(DummyInstructions),
-      // SEC_CM: DATA_REG_SW.GLITCH_DETECT
-      .WrenCheck        (RegFileWrenCheck),
-      .RdataMuxCheck    (RegFileRdataMuxCheck),
       .WordZeroVal      (RegFileDataWidth'(prim_secded_pkg::SecdedInv3932ZeroWord))
     ) register_file_i (
       .clk_i (clk),
@@ -440,17 +434,13 @@ module ibex_top import ibex_pkg::*; #(
       .rdata_b_o(rf_rdata_b_ecc),
       .waddr_a_i(rf_waddr_wb),
       .wdata_a_i(rf_wdata_wb_ecc),
-      .we_a_i   (rf_we_wb),
-      .err_o    (rf_alert_major_internal)
+      .we_a_i   (rf_we_wb)
     );
   end else if (RegFile == RegFileFPGA) begin : gen_regfile_fpga
     ibex_register_file_fpga #(
       .RV32E            (RV32E),
       .DataWidth        (RegFileDataWidth),
       .DummyInstructions(DummyInstructions),
-      // SEC_CM: DATA_REG_SW.GLITCH_DETECT
-      .WrenCheck        (RegFileWrenCheck),
-      .RdataMuxCheck    (RegFileRdataMuxCheck),
       .WordZeroVal      (RegFileDataWidth'(prim_secded_pkg::SecdedInv3932ZeroWord))
     ) register_file_i (
       .clk_i (clk),
@@ -466,17 +456,13 @@ module ibex_top import ibex_pkg::*; #(
       .rdata_b_o(rf_rdata_b_ecc),
       .waddr_a_i(rf_waddr_wb),
       .wdata_a_i(rf_wdata_wb_ecc),
-      .we_a_i   (rf_we_wb),
-      .err_o    (rf_alert_major_internal)
+      .we_a_i   (rf_we_wb)
     );
   end else if (RegFile == RegFileLatch) begin : gen_regfile_latch
     ibex_register_file_latch #(
       .RV32E            (RV32E),
       .DataWidth        (RegFileDataWidth),
       .DummyInstructions(DummyInstructions),
-      // SEC_CM: DATA_REG_SW.GLITCH_DETECT
-      .WrenCheck        (RegFileWrenCheck),
-      .RdataMuxCheck    (RegFileRdataMuxCheck),
       .WordZeroVal      (RegFileDataWidth'(prim_secded_pkg::SecdedInv3932ZeroWord))
     ) register_file_i (
       .clk_i (clk),
@@ -492,8 +478,7 @@ module ibex_top import ibex_pkg::*; #(
       .rdata_b_o(rf_rdata_b_ecc),
       .waddr_a_i(rf_waddr_wb),
       .wdata_a_i(rf_wdata_wb_ecc),
-      .we_a_i   (rf_we_wb),
-      .err_o    (rf_alert_major_internal)
+      .we_a_i   (rf_we_wb)
     );
   end
 
@@ -1106,7 +1091,6 @@ module ibex_top import ibex_pkg::*; #(
 
   assign alert_major_internal_o = core_alert_major_internal |
                                   lockstep_alert_major_internal |
-                                  rf_alert_major_internal |
                                   icache_alert_major_internal;
   assign alert_major_bus_o      = core_alert_major_bus | lockstep_alert_major_bus;
   assign alert_minor_o          = core_alert_minor | lockstep_alert_minor;
