diff --git b/BUILD a/BUILD
new file mode 100644
index 000000000..7a3961010
--- /dev/null
+++ a/BUILD
@@ -0,0 +1,52 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+package(default_visibility = ["//visibility:public"])
+
+load("//rules:opentitan.bzl", "OPENTITAN_CPU")
+
+cc_library(
+    name = "portmacro",
+    includes = ["portable/GCC/RISC-V"],
+    hdrs = [
+        "portable/GCC/RISC-V/portmacro.h",
+    ],
+    target_compatible_with = [OPENTITAN_CPU],
+)
+
+cc_library(
+    name = "kernel",
+    srcs = [
+        "tasks.c",
+        "queue.c",
+        "list.c",
+        "portable/MemMang/heap_1.c",
+        "include/deprecated_definitions.h",
+    ],
+    includes = ["include"],
+    hdrs = [
+        "include/atomic.h",
+        "include/croutine.h",
+        "include/event_groups.h",
+        "include/FreeRTOS.h",
+        "include/list.h",
+        "include/message_buffer.h",
+        "include/mpu_prototypes.h",
+        "include/mpu_wrappers.h",
+        "include/portable.h",
+        "include/projdefs.h",
+        "include/queue.h",
+        "include/semphr.h",
+        "include/stack_macros.h",
+        "include/StackMacros.h",
+        "include/stream_buffer.h",
+        "include/task.h",
+        "include/timers.h",
+    ],
+    deps = [
+        ":portmacro",
+        "//sw/device/lib/base",
+        "//sw/device/lib/testing/test_framework:freertos_config",
+    ],
+)
