diff --git a/fuzzing/init.bzl b/fuzzing/init.bzl
index f8149d8..1823be7 100644
--- a/fuzzing/init.bzl
+++ b/fuzzing/init.bzl
@@ -20,7 +20,6 @@ load("@rules_python//python:pip.bzl", "pip_parse")
 def rules_fuzzing_init():
     pip_parse(
         name = "fuzzing_py_deps",
-        extra_pip_args = ["--require-hashes"],
         requirements_lock = "@rules_fuzzing//fuzzing:requirements.txt",
     )
     bazel_skylib_workspace()
